RationaleMCP/0027/HindleyMilner.md

```diff
@@ -0,0 +1,268 @@
+# Unit checking
+
+Unit checking is the process of inferring the units of Modelica variables, and ensuring the variables are used consistently in expressions and equations.
+
+It takes place after any evaluation of evaluable parameters and no evaluable parameters shall be evaluated during this process.
+(This must be true as long as tools allow the users to opt-out from unit-checking.)
+
+It is performed by using a type inference algorithm, albeit adapted to work with units.
+The algorithm proposed here is based on the Hindley-Milner algorithm.
+At a high level the algorithm collects a set of constraints on the units of expressions.
+It extracts a sequence of substitutions from these constraints which once applied constructs the unit of a given expression.
```

Comment:
I don't understand what _a given expression_ is supposed to mean here.  I even wonder if it should be highlighted at all that we can use the sequence to find units of expressions, at least not until we get to things like `inUnit`.

Maybe this would be more to the point, at least for now?
```suggestion
The constraints are solved by iteratively eliminating the unknowns of the problem using substitutions derived from the constraints.
```

Reply:
Not all unknowns are eliminated though.

Reply:
I am trying to describe how one goes from the sequence of substitution to the unit of a variable, by highlighting that applying the sequence of substitution to an arbitrary unit meta-expression, will produce a well-formed unit, if it can be inferred.